### ISSUE TITLE
Proposal for new module: shortcuts

### ISSUE THREAD

**Issue by poiru** [electron/electron#20964]
Right now, we have the `globalShortcuts` module for registering shortcuts that apply to the entire system even when the Electron app is not focused.

I'd like to propose a new `shortcuts` module and deprecating the existing `globalShortcuts` module. The API would look like so:

```js
const { shortcut } = require('electron');

let alreadyRegistered = shortcut.register('normal', 'CmdOrCtrl+L', () => {
  // Normal priority shortcut, handled after web content. Web content may preventDefault().
});

let alreadyRegistered = shortcut.register('high', 'CmdOrCtrl+L', () => {
  // High priority shortcut, handled before web content.
});

let alreadyRegistered = shortcut.register('global', 'CmdOrCtrl+L', () => {
  // Similar to existing globalShortcuts.register(), handled before web content.
});

let registed = shortcut.isRegistered('CmdOrCtrl+L');

shortcut.unregister('CmdOrCtrl+L');
```

There is clear demand for this feature, see #1334 and https://github.com/parro-it/electron-localshortcut

Implementing this within Electron would be much better because the accelerators would be parsed and handled like we already handle `Menu` accelerators. Properly processing the accelerators is tricky and error-prone - `electron-localshortcut` has several bugs and we have had multiple bugs in our own custom implementation as well.

I'm happy to implement this, but I wanted to hear your thoughts first!

cc @electron/wg-api

**Comment by Kilian** [comment]
Would this also be a place to look at https://github.com/electron/electron/issues/19747 ?

In-webContent shortcuts (using keypress) work correctly since they use chromium's key handling, but they only work with the webContents focused and have no support for e.g. function keys like the current globalShortcuts do.

**Comment by JamesCoyle** [comment]
I feel a syntax such as the following would be nicer: 

```js
let shortcut = shortcut.register('CmdOrCtrl+L', () => {
  // Normal priority shortcut, handled after web content. Web content may preventDefault().
});

let shortcutIntercept = shortcut.register('CmdOrCtrl+L', () => {
  // High priority shortcut, handled before web content.
}, true);

let shortcutGlobal = shortcut.registerGlobal('CmdOrCtrl+L', () => {
  // Similar to existing globalShortcuts.register(), handled before web content.
});
```

**Comment by sindresorhus** [comment]
I think this would be the most readable:

```js
let shortcut = shortcut.registerLocal('CmdOrCtrl+L', () => {
  // Normal priority shortcut, handled after web content. Web content may preventDefault().
});

let shortcutIntercept = shortcut.registerLocal('CmdOrCtrl+L', () => {
  // High priority shortcut, handled before web content.
}, {
  highPriority: true
});

let shortcutGlobal = shortcut.registerGlobal('CmdOrCtrl+L', () => {
  // Similar to existing globalShortcuts.register(), handled before web content.
});
```

It's now clear what they do even when `registerLocal` and `registerGlobal` are alone what they do exactly. I would also recommend using an options object over a [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap).

---

Another idea would be to make the shortcut itself a type. This would make it nicer to compare shortcuts.

```js
const fooShortcut = shortcut('CmdOrCtrl+L');

fooShortcut.registerLocal(() => {
  // Normal priority shortcut, handled after web content. Web content may preventDefault().
});

fooShortcut.registerLocal(() => {
  // High priority shortcut, handled before web content.
}, {
  highPriority: true
});

fooShortcut.registerGlobal(() => {
  // Similar to existing globalShortcuts.register(), handled before web content.
});

fooShortcut.isRegistered;

fooShortcut.unregister();

fooShortcut === shortcut('CmdOrCtrl+L');
//=> true
```

**Comment by fabiospampinato** [comment]
Maybe it would be more appropriate and web-friendly to use a third-party library for this, like this one: https://github.com/fabiospampinato/shortcuts

Like what feature could Electron offer than third-party libraries can't offer too? And if this is implementable completely via third-party libraries why going for an Electron-only version which would  be painful to move from once/if the app will also ever run in the browser.

**Comment by mathieulj** [comment]
@fabiospampinato Maybe I misunderstood something here but aren't the libraries you mentioned only able to capture events when focus is within the web view? `globalShortcuts` is specifically for handling keyboard shortcuts regardless of if or not the application has focus and it is meant to work even with the application minimized.

**Comment by fabiospampinato** [comment]
@mathieulj precisely, but `globalShortcuts` we have already, the proposal in this issue is to add support for capturing local shortcuts too.

**Comment by codebytere** [comment]
Hi, while we appreciate this enhancement request it's not something that our team will be able to implement in the near future, and since we use our issues tab as the team's backlog, we prefer to only retain issues that are on our immediate roadmap. If you'd like to make a PR for this request we would welcome that but in the meantime I'm going to go ahead and close this. 

Further proposal work should be directed to https://github.com/electron/rfcs
